libbpf-tools/execsnoop.bpf.c

```diff
@@ -0,0 +1,129 @@
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include "execsnoop.h"
+
+// null isn't defined
+#define NULL 0
+
+const volatile bool ignore_failed = true;
+const volatile uid_t targ_uid = INVALID_UID;
+const volatile int max_args = DEFAULT_MAXARGS;
+
+static const struct event empty_event = {};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10240);
+	__type(key, pid_t);
+	__type(value, struct event);
+} execs SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+static __always_inline bool valid_uid(uid_t uid) {
+	return uid != INVALID_UID;
+}
+
+SEC("tracepoint/syscalls/sys_enter_execve")
+int tracepoint__syscalls__sys_enter_execve(struct trace_event_raw_sys_enter* ctx)
+{
+	u64 id;
+	pid_t pid, tgid;
+	int ret;
+	struct event *event;
+	uid_t uid = (u32)bpf_get_current_uid_gid();
+	if (valid_uid(targ_uid) && targ_uid != uid)
+		return 0;
+
+	id = bpf_get_current_pid_tgid();
+	pid = (pid_t)id;
+	tgid = id >> 32;
+	if (bpf_map_update_elem(&execs, &pid, &empty_event, BPF_NOEXIST))
+		return 0;
+
+	event = bpf_map_lookup_elem(&execs, &pid);
+	if (!event)
+		return 0;
+
+	event->pid = pid;
+	event->tgid = tgid;
+	event->uid = uid;
+	event->more_args = false;
+	event->args_count = 0;
+	event->args_size = 0;
+
+	ret = bpf_probe_read_str(event->args, ARGSIZE, (const char*)ctx->args[0]);
+	if (ret > 0 && ret <= ARGSIZE) {
+		event->args_size += ret;
+	} else {
+		/* write an empty string */
+		event->args[0] = '\0';
+		event->args_size++;
+	}
+
+	event->args_count++;
+	const char **args = (const char **)(ctx->args[1]);
+	for (int i = 1; i < max_args; i++) {
+		const char *argp = NULL;
+
+		bpf_probe_read(&argp, sizeof(argp), &args[i]);
+
+		if (!argp)
+			return 0;
+
+		if (event->args_size < 0 || event->args_size > ARGS_END)
```

Comment:
same as suggested above, make args_size unsigned and only check against ARGS_END


Also I think ARGS_END should equal to FULL_MAX_ARGS_ARR, because `i` starts at 1 and `ctx->args[0]` is treated as a first argument (and is counted against same max_args limit). This might be why 5.4 verifier is complaining.

Reply:
`Also I think ARGS_END should equal to FULL_MAX_ARGS_ARR,`

This would be incorrect and in that case verifier reject code. `ARGS_END` is really a wrong name for such constant, it's actually an offset that has at enough space for at least one argument (there is call `bpf_probe_read_str` after this check passes).

Reply:
Ah, right, yeah, the name is definitely misleading. You could do `event->args_size + ARGSIZE > ARGS_END` and compiler should optimize it to same thing, I think.